### PR TITLE
New resource: aws_apigatewayv2_api_mapping

### PR DESCRIPTION
**Pull request by ewbankkit**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Another resource for #7004.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
FEATURES:

* New Resource: aws_api_gateway_v2_api_mapping (#7004)
```

Output from acceptance testing:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSAPIGateway2ApiMapping_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAWSAPIGateway2ApiMapping_ -timeout 120m
=== RUN   TestAccAWSAPIGateway2ApiMapping_basic
=== PAUSE TestAccAWSAPIGateway2ApiMapping_basic
=== RUN   TestAccAWSAPIGateway2ApiMapping_ApiMappingKey
=== PAUSE TestAccAWSAPIGateway2ApiMapping_ApiMappingKey
=== CONT  TestAccAWSAPIGateway2ApiMapping_basic
=== CONT  TestAccAWSAPIGateway2ApiMapping_ApiMappingKey
--- PASS: TestAccAWSAPIGateway2ApiMapping_basic (34.49s)
--- PASS: TestAccAWSAPIGateway2ApiMapping_ApiMappingKey (97.05s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	97.120s
```

Currently based on https://github.com/terraform-providers/terraform-provider-aws/pull/8842, will need to rebase once that PR is merged.

### CODE REVIEW

**Comment by ewbankkit** [comment]
Includes
* https://github.com/terraform-providers/terraform-provider-aws/pull/9391 for `aws_api_gateway_v2_domain_name` tests
* https://github.com/terraform-providers/terraform-provider-aws/pull/9232 for `aws_api_gateway_v2_stage` tests

**Comment by ewbankkit** [comment]
Rebased to fix merge conflict.
Acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSAPIGateway2ApiMapping_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAWSAPIGateway2ApiMapping_ -timeout 120m
=== RUN   TestAccAWSAPIGateway2ApiMapping_basic
=== PAUSE TestAccAWSAPIGateway2ApiMapping_basic
=== RUN   TestAccAWSAPIGateway2ApiMapping_ApiMappingKey
=== PAUSE TestAccAWSAPIGateway2ApiMapping_ApiMappingKey
=== CONT  TestAccAWSAPIGateway2ApiMapping_basic
=== CONT  TestAccAWSAPIGateway2ApiMapping_ApiMappingKey
--- PASS: TestAccAWSAPIGateway2ApiMapping_ApiMappingKey (52.60s)
--- PASS: TestAccAWSAPIGateway2ApiMapping_basic (72.66s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	72.707s
```

File: aws/structure.go [hunk 1]
```diff
@@ -1477,6 +1477,7 @@ func flattenConfigSnapshotDeliveryProperties(p *configservice.ConfigSnapshotDeli
 	return []map[string]interface{}{m}
 }
 
+// TODO This should really be named 'pointersMapToStringMap'?
```
Comment (haavardw) [inline aws/structure.go:1480]: Intentional to leave these comments? (seems unrelated to this PR?)

Comment (ewbankkit) [reply aws/structure.go:1480]: Indeed, it must have just been left over from when I was working on the code.
Once some more of the prerequisite PRs for "lower" API Gateway v2 APIs have been merged I will rebase this one and remove the comments (and maybe revisit `diffStringMaps` - the intervening work on `keyvaluetags` may have some code I can reuse).

**Comment by ewbankkit** [comment]
After merge of https://github.com/terraform-providers/terraform-provider-aws/pull/9391 and https://github.com/terraform-providers/terraform-provider-aws/pull/9232, incorporate all the feedback:
* Migrate to Terraform provider SDK
* Change resource name to `aws_apigatewayv2_blah`; Rename source files
* Remove `sidebar_current` from documentation
* Add `subcategory: "API Gateway v2 (WebSocket and HTTP APIs)"` to documentation
* Change function names to `ApiGatewayV2`
* Add `_disappears` acceptance test
* Use `multierror` in any test sweeper

**Comment by ewbankkit** [comment]
Renamed resource to `aws_apigatewayv2_api_mapping`.
Re-ran acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSAPIGatewayV2ApiMapping'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSAPIGatewayV2ApiMapping -timeout 120m
=== RUN   TestAccAWSAPIGatewayV2ApiMapping
=== RUN   TestAccAWSAPIGatewayV2ApiMapping/createCertificate
=== RUN   TestAccAWSAPIGatewayV2ApiMapping/basic
=== RUN   TestAccAWSAPIGatewayV2ApiMapping/disappears
=== RUN   TestAccAWSAPIGatewayV2ApiMapping/ApiMappingKey
--- PASS: TestAccAWSAPIGatewayV2ApiMapping (118.36s)
    --- PASS: TestAccAWSAPIGatewayV2ApiMapping/createCertificate (3.25s)
    --- PASS: TestAccAWSAPIGatewayV2ApiMapping/basic (32.32s)
    --- PASS: TestAccAWSAPIGatewayV2ApiMapping/disappears (30.17s)
    --- PASS: TestAccAWSAPIGatewayV2ApiMapping/ApiMappingKey (52.61s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	118.433s
```

Closes #7004.

**Comment by ewbankkit** [comment]
The acceptance tests require that an ACM certificate be created outside of Terraform configuration (and left dangling after the tests finish) as any certificate created as part of the test configuration is in an associated state when deletion is attempted. This seems to be because the API Gateway domain name has a backing CloudFront distribution (in an AWS-owned account) and that can take up to 60 minutes to delete:
* https://docs.aws.amazon.com/acm/latest/userguide/troubleshoot-apigateway.html
* https://aws.amazon.com/premiumsupport/knowledge-center/acm-certificate-resources/

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSAPIGatewayV2ApiMapping_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -count 1 -parallel 20 -run=TestAccAWSAPIGatewayV2ApiMapping_ -timeout 120m
=== RUN   TestAccAWSAPIGatewayV2ApiMapping_basic
=== PAUSE TestAccAWSAPIGatewayV2ApiMapping_basic
=== RUN   TestAccAWSAPIGatewayV2ApiMapping_ApiMappingKey
=== PAUSE TestAccAWSAPIGatewayV2ApiMapping_ApiMappingKey
=== CONT  TestAccAWSAPIGatewayV2ApiMapping_basic
=== CONT  TestAccAWSAPIGatewayV2ApiMapping_ApiMappingKey
--- FAIL: TestAccAWSAPIGatewayV2ApiMapping_basic (635.71s)
    testing.go:730: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.
        
        Error: errors during apply: Error deleting certificate: ResourceInUseException: Certificate arn:aws:acm:us-west-2:123456789012:certificate/e2e672f5-1d17-4338-bd8d-98881d622293 in account 123456789012 is in use.
        
        State: aws_acm_certificate.test:
          ID = arn:aws:acm:us-west-2:123456789012:certificate/e2e672f5-1d17-4338-bd8d-98881d622293
          provider = provider.aws
          arn = arn:aws:acm:us-west-2:123456789012:certificate/e2e672f5-1d17-4338-bd8d-98881d622293
          certificate_authority_arn = 
          certificate_body = 790e1bd1153e851211df8beaf27cd09ecb1d21bc
          domain_name = tf-acc-test-7526054328147068447.example.com
          domain_validation_options.# = 0
          options.# = 1
          options.0.certificate_transparency_logging_preference = DISABLED
          private_key = 21d2128df77fad2a11f6d8b94470538213c74f12
          subject_alternative_names.# = 0
          tags.% = 1
          tags.Name = tf-acc-test-7526054328147068447
          validation_emails.# = 0
          validation_method = NONE
--- FAIL: TestAccAWSAPIGatewayV2ApiMapping_ApiMappingKey (668.35s)
    testing.go:730: Error destroying resource! WARNING: Dangling resources
        may exist. The full state and error is shown below.
        
        Error: errors during apply: Error deleting certificate: ResourceInUseException: Certificate arn:aws:acm:us-west-2:123456789012:certificate/59441bc8-10e1-4679-848b-3996e65b7bbf in account 123456789012 is in use.
        
        State: aws_acm_certificate.test:
          ID = arn:aws:acm:us-west-2:123456789012:certificate/59441bc8-10e1-4679-848b-3996e65b7bbf
          provider = provider.aws
          arn = arn:aws:acm:us-west-2:123456789012:certificate/59441bc8-10e1-4679-848b-3996e65b7bbf
          certificate_authority_arn = 
          certificate_body = a245b90590f966072471d91eb7d54c0d47d11185
          domain_name = tf-acc-test-7248658647982118682.example.com
          domain_validation_options.# = 0
          options.# = 1
          options.0.certificate_transparency_logging_preference = DISABLED
          private_key = cc95ddca77f902c828ef60aa931cb55b3b6614da
          subject_alternative_names.# = 0
          tags.% = 1
          tags.Name = tf-acc-test-7248658647982118682
          validation_emails.# = 0
          validation_method = NONE
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	668.512s
FAIL
GNUmakefile:26: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

**Comment by ewbankkit** [comment]
Ready for review.

File: aws/resource_aws_apigatewayv2_api_mapping_test.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/acm"
+	"github.com/aws/aws-sdk-go/service/apigatewayv2"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/terraform"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags"
+)
+
```
Comment (gdavison) [inline aws/resource_aws_apigatewayv2_api_mapping_test.go:16]: Should this have a sweeper, or will the resources get cleaned up when other resources are deleted?

Comment (ewbankkit) [reply aws/resource_aws_apigatewayv2_api_mapping_test.go:16]: It's the `aws_acm_certificate` that was created that needs sweeping - I'll add a sweeper as that resource doesn't have one currently.
If I try to sweep APIs with API mappings defined I get:

```console
$ SWEEP=us-west-2 SWEEP_DIR=./aws/ SWEEPARGS=-sweep-run=aws_apigatewayv2_api make sweep
WARNING: This will destroy infrastructure. Use only in development accounts.
go test ./aws/ -v -sweep=us-west-2 -sweep-run=aws_apigatewayv2_api -timeout 60m
2020/04/18 17:56:38 [DEBUG] Running Sweepers for region (us-west-2):
2020/04/18 17:56:38 [DEBUG] Running Sweeper (aws_apigatewayv2_api) in region (us-west-2)
2020/04/18 17:56:38 [INFO] Building AWS auth structure
2020/04/18 17:56:38 [INFO] Setting AWS metadata API timeout to 100ms
2020/04/18 17:56:39 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
2020/04/18 17:56:39 [INFO] AWS Auth provider used: "EnvProvider"
2020/04/18 17:56:39 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/04/18 17:56:40 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/04/18 17:56:41 [INFO] Deleting API Gateway v2 API: bvowp2c4b5
2020/04/18 17:56:41 [ERROR] error deleting API Gateway v2 API (bvowp2c4b5): BadRequestException: Deleting RestApi bvowp2c4b5 failed. Please remove all base path mappings related to the RestApi in your domains.
2020/04/18 17:56:41 [ERROR] Error running Sweeper (aws_apigatewayv2_api) in region (us-west-2): 1 error occurred:
	* error deleting API Gateway v2 API (bvowp2c4b5): BadRequestException: Deleting RestApi bvowp2c4b5 failed. Please remove all base path mappings related to the RestApi in your domains.

FAIL	github.com/terraform-providers/terraform-provider-aws/aws	3.718s
FAIL
GNUmakefile:19: recipe for target 'sweep' failed
make: *** [sweep] Error 1
```

That error message seems to be a holdover from API Gateway v1 but it means that domain names need to be deleted first, so I've added a dependency to the `aws_apigatewayv2_api` sweeper:

```console
$ SWEEP=us-west-2 SWEEP_DIR=./aws/ SWEEPARGS=-sweep-run=aws_apigatewayv2_api make sweep
WARNING: This will destroy infrastructure. Use only in development accounts.
go test ./aws/ -v -sweep=us-west-2 -sweep-run=aws_apigatewayv2_api -timeout 60m
2020/04/18 18:14:38 [DEBUG] Running Sweepers for region (us-west-2):
2020/04/18 18:14:38 [DEBUG] Sweeper (aws_apigatewayv2_api) has dependency (aws_apigatewayv2_domain_name), running..
2020/04/18 18:14:38 [DEBUG] Running Sweeper (aws_apigatewayv2_domain_name) in region (us-west-2)
2020/04/18 18:14:38 [INFO] Building AWS auth structure
2020/04/18 18:14:38 [INFO] Setting AWS metadata API timeout to 100ms
2020/04/18 18:14:40 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
2020/04/18 18:14:40 [INFO] AWS Auth provider used: "EnvProvider"
2020/04/18 18:14:40 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/04/18 18:14:40 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/04/18 18:14:41 [INFO] Deleting API Gateway v2 domain name: example-ws-api.example.com
2020/04/18 18:14:42 [DEBUG] Running Sweeper (aws_apigatewayv2_api) in region (us-west-2)
2020/04/18 18:14:42 [INFO] Building AWS auth structure
2020/04/18 18:14:42 [INFO] Setting AWS metadata API timeout to 100ms
2020/04/18 18:14:43 [INFO] Ignoring AWS metadata API endpoint at default location as it doesn't return any instance-id
2020/04/18 18:14:43 [INFO] AWS Auth provider used: "EnvProvider"
2020/04/18 18:14:43 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/04/18 18:14:43 [DEBUG] Trying to get account information via sts:GetCallerIdentity
2020/04/18 18:14:44 [INFO] Deleting API Gateway v2 API: m36pbck0ii
2020/04/18 18:14:46 [DEBUG] Sweeper (aws_apigatewayv2_domain_name) already ran in region (us-west-2)
2020/04/18 18:14:46 Sweeper Tests ran successfully:
	- aws_apigatewayv2_domain_name
	- aws_apigatewayv2_api
ok  	github.com/terraform-providers/terraform-provider-aws/aws	7.842s
```
